### PR TITLE
test run of module conversion to es6

### PR DESCRIPTION
**Pull request by sndrs**
## What does this change?

- converts `projects/common/utils/$.js` to es6
- lints unnecessary lodash array methods

## What is the value of this and can you measure success?

see the conversion script performs, and whether es6 env is setup correctly

## Does this affect other platforms - Amp, Apps, etc?

no

## Tested in CODE?

- [x] test on CODE

### CODE REVIEW

File: static/src/javascripts/.eslintrc.js [hunk 1]
```diff
@@ -31,7 +31,12 @@ module.exports = {
         'no-restricted-imports': [
             'error',
             {
-                paths: ['lodash'],
+                paths: [
+                    'lodash',
+                    'lodash/collections/forEach',
+                    'lodash/collections/map',
+                    'lodash/collections/reduce',
+                ],
```
Comment (gustavpursche) [inline static/src/javascripts/.eslintrc.js]: What do you think about adding 

- [ ] `collections/filter`
- [ ] `collections/some`
- [ ] `objects/assign`
- [ ] `objects/values`
- [ ] `objects/merge`
- [ ] `objects/keys`
- [ ] `collections/every`
- [ ] `collections/contains`
- [ ] `objects/isArray`
- [ ] `arrays/indexOf`

This would allow the removal a huge part of lodash ...

File: static/src/javascripts/projects/common/utils/$.js [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+        return el.parentNode;
+    }
+    return $.ancestor(el.parentNode, c);
+};
+
+$.forEachElement = (selector, fn) => {
+    const els = qwery(selector);
+    els.forEach(fn);
+    return els;
+};
```
Comment (gustavpursche) [inline static/src/javascripts/projects/common/utils/$.js]: Wouldn't it be better that way:

```js
$.forEachElement = (selector, fn) => {
    const els = qwery(selector);
    return els.map(fn);
};
```

File: static/src/javascripts/.eslintrc.js (hunk 1 above)
Comment (sndrs) [inline static/src/javascripts/.eslintrc.js]: 💯 these came off the top of my head, but totally. will update

File: static/src/javascripts/projects/common/utils/$.js (hunk 2 above)
Comment (sndrs) [inline static/src/javascripts/projects/common/utils/$.js]: ha indeed it would! so even converting the simplest modules we musn't be lazy...good one

Comment (gustavpursche) [inline static/src/javascripts/projects/common/utils/$.js]: Does it return a NodeList instead?

Comment (sndrs) [inline static/src/javascripts/projects/common/utils/$.js]: ooh yeah you're correct, i misread the docs, thought it returned an enhanced array-like object, but doesn't look like it all. will test it in the morning, but looks good!

Comment (desbo) [inline static/src/javascripts/projects/common/utils/$.js]: simplification looks good but I'd stick with `forEach` rather than `map` – you wouldn't expect a return value from a function called `forEachElement` so there's no need to build an array.

Comment (sndrs) [inline static/src/javascripts/projects/common/utils/$.js]: it did already though?

Comment (desbo) [inline static/src/javascripts/projects/common/utils/$.js]: true, but none of the current uses of it do anything with the returned value. could be a good optimisation ([jsperf](https://jsperf.com/foreach-map-speed/1)) unless there's a reason i'm missing for returning the nodes

Comment (gustavpursche) [inline static/src/javascripts/projects/common/utils/$.js]: I just checked, how often `forEachElement` is used. Turns out: 7 times, so it's not as critical as I thought.

1. What do you think about taking it out of `utils/$` and move it to it's own utility, to clearify, it's not part of `bonzo`?

2. I'm not very opinionated about whether to use `map` for `forEach` in this case. I'm asking myself, why it exists anywy, since there is [bonzo().each()](https://github.com/ded/bonzo#api-each).

3. 🐎 Let's not fall into a micro optimization trap for now - just make `forEachElement` work - my fault.

Comment (sndrs) [inline static/src/javascripts/projects/common/utils/$.js]: it does throw up an interesting question with regards to how much we want to revisit code as we convert it. i guess that would have to be up to the individual, since it's their PR and their responsibility to deal with any errors it introduces etc. 

there's some good points in this conversation though, but i wonder if they would be better addressed as separate PRs?

Comment (gustavpursche) [inline static/src/javascripts/projects/common/utils/$.js]: > there's some good points in this conversation though, but i wonder if they would be better addressed as separate PRs?

👍 

Does this work though?

```js
const els = qwery(selector);
els.forEach(fn);
return els;
```

I'd give you my approval then, so we can move on :)

**Review by SiAdcock — APPROVED** [review]

**Review by NataliaLKB — APPROVED** [review]

**Comment by sndrs** [comment]
closing while we look at utils in more depth (sorry everyone!)
